src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -79,6 +79,75 @@ - (BOOL)callDelegateSynchronouslyWithBlock:(void (^)(id<MTRDeviceDelegate>))bloc
 #endif
 @end
 
+#pragma mark - MTRAttributeValueWaiter
+
+static NSString * const MTRValueSatisfiedKey = @"valueSatisfied";
+
+// Represents the data passed to waitForAttributeValues:
+@interface MTRAttributeValueWaiter : NSObject
+// Each entry of "valueExpectations" has two keys:
+// * MTRValueSatisfiedKey: whether that value has been satisfied by observed
+//                         attribute values.
+// * MTRValueKey: the response-value we are trying to wait for.
+@property (nonatomic, retain) NSArray<NSMutableDictionary<NSString *, id> *> * valueExpectations;
+@property (nonatomic, retain) dispatch_queue_t queue;
+@property (nonatomic) MTRStatusCompletion completion;
+
+@property (nonatomic, readonly) BOOL allValuesSatisfied;
+
+- (instancetype)initWithValues:(NSArray<NSDictionary<NSString *, id> *> *)values queue:(dispatch_queue_t)queue completion:(MTRStatusCompletion)completion;
+
+// Returns YES if after this report the waiter might be done waiting, NO otherwise.
+- (BOOL)attributeValue:(MTRDeviceDataValueDictionary)value reportedForPath:(MTRAttributePath *)path byDevice:(MTRDevice *)device;
+
+@end
+
+@implementation MTRAttributeValueWaiter
+- (instancetype)initWithValues:(NSArray<NSDictionary<NSString *, id> *> *)values queue:(dispatch_queue_t)queue completion:(MTRStatusCompletion)completion
+{
+    if (self = [super init]) {
+        auto * valueExpectations = [NSMutableArray arrayWithCapacity:values.count];
+        for (NSDictionary<NSString *, id> * value in values) {
+            auto * valueExpectation = [NSMutableDictionary dictionaryWithCapacity:2];
```

Comment:
Can we not use a class for this instead of a dictionary?

Reply:
We could, sure.  Wasn't sure it was worth creating another class, but I could add one.

Reply:
Done.